apm-lambda-extension/main.go

```diff
@@ -121,17 +122,17 @@ func main() {
 			// Make a channel for signaling that the function invocation is complete
 			funcDone := make(chan struct{})
 
-			// Flush any APM data, in case waiting for the agentDone or runtimeDone signals
-			// timed out, the agent data wasn't available yet, and we got to the next event
-			extension.FlushAPMData(client, agentDataChannel, config)
-
 			// A shutdown event indicates the execution environment is shutting down.
 			// This is usually due to inactivity.
 			if event.EventType == extension.Shutdown {
 				extension.ProcessShutdown()
 				return
 			}
 
+			// Flush any APM data, in case waiting for the agentDone or runtimeDone signals
+			// timed out, the agent data wasn't available yet, and we got to the next non-shutdown event
+			extension.FlushAPMData(client, agentDataChannel, config)
+
```

Comment:
Why did you move this block to after the Shutdown event check?

Reply:
When writing my tests, I found myself in the following situation : 

Test setup : 
- APM Server configured to be very slow (approx 3 seconds to respond).
- APM Agent Data channel filled with data
- SendStrategy configured to background

Behavior :
The extension starts sending data from the loaded channel to the slow APM server. It is then interrupted due to the current event being finished. A shutdown event is then received, and we start flushing APM data. The issue is that this takes a tremendous time if the APM Agent data channel is full and if the response time of the Server is large, even if the SendStrategy is set to `background`.

An alternative solution would be to put the flush statement above the Shutdown, but to only run `FlushAPMData` if the SendStrategy is `SyncFlush`

Reply:
Thanks for the explanation. I think we might want to remove this flush anyway: with `syncFlush`, we'll flush at the end of an invocation so there won't be anything to flush upon a new function invocation. With `background`, we'll just continue receiving from the channel in the go routine when the extension starts its work for a new function invocation.